sdk/iot/pnp/inc/az_iot_pnp_client.h

```diff
@@ -0,0 +1,334 @@
+ *                     policy key.
+ * @param[in] mqtt_password An empty #az_span with sufficient capacity to hold the MQTT password.
+ * @param[out] out_mqtt_password The output #az_span containing the MQTT password.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_sas_password_get(
+    az_iot_pnp_client const* client,
+    az_span base64_hmac_sha256_signature,
+    az_span key_name,
+    az_span mqtt_password,
+    az_span* out_mqtt_password);
+
+/*
+ *
+ * PnP Telemetry APIs
+ *
+ */
+
+/**
+ * @brief Azure IoT PnP telemetry options.
+ *
+ */
+typedef struct az_iot_pnp_client_telemetry_options
+{
+  struct
+  {
+    // OPEN: I can't find samples in Azure SDK modeling this.  Closest is
+    // sdk\storage\blobs\inc\az_storage_blobs.h where it has 'az_span option" in
+    // az_storage_blobs_blob_download_options that's ignored in practice.
+    int _not_used;
+  } _internal;
+} az_iot_pnp_client_telemetry_options;
+
+/**
+ * @brief Gets the default Azure IoT PnP telemetry options.
+ * @details Call this to obtain an initialized #az_iot_pnp_client_telemetry_options structure that
+ * can be afterwards modified and passed to #az_iot_pnp_client_telemetry_publish_topic_get.
+ *
+ * @return #az_iot_pnp_client_telemetry_options.
+ */
+AZ_NODISCARD az_iot_pnp_client_telemetry_options az_iot_pnp_client_telemetry_options_default();
+
+/**
+ * @brief Gets the MQTT topic that must be used for device to cloud telemetry messages.
+ * @note Telemetry MQTT Publish messages must have QoS At Least Once (1).
+ * @note This topic can also be used to set the MQTT Will message in the Connect message.
+ *
+ * Should the user want a null terminated topic string, they may allocate a buffer large enough
+ * to fit the topic plus a null terminator. They must set the last byte themselves or zero
+ * initialize the buffer.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] component_name An #az_span specifying the component name to publish telemetry on.
+ * @param[in] content_type An #az_span specifying the content type of the message payload.
+                           Only AZ_SPAN_FROM_STR("application/json") is currently supported.
+ * @param[in] encoding_type An #az_span specifying the encoding type of the message payload.
+                           Only AZ_SPAN_FROM_STR("utf-8") is currently supported.
+ * @param[in] options A reference to an #az_iot_pnp_client_telemetry_options structure. Can be NULL.
+ * @param[in] mqtt_topic An empty #az_span with sufficient capacity to hold the MQTT topic.
+ * @param[out] out_mqtt_topic The output #az_span containing the MQTT topic.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_telemetry_publish_topic_get(
+    az_iot_pnp_client const* client,
+    az_span component_name,
+    az_span content_type,
+    az_span encoding_type,
+    az_iot_pnp_client_telemetry_options const* options,
```

Comment:
Same here and everywhere else (if you decide to keep), it should probably be the last argument.

Reply:
Tough call on this one. I feel that output parameters should be last so maybe just before out_mqtt_topic?
Although I know it';s nice to have mqtt_topic next to out_mqtt_topic.
I don't have an answer. I'm open to suggestions.

Reply:
I personally like the ordering being (client_handle, in-required1, ... in-required<n>, optional, &out_param).  That's how I've seen it in other projects at least.  It's unfortunate that the out_mqtt_topic gets split from its buddy though, but I don't like optional in middle of in-required*, either.

I'm going to go with this *for now*.  As this is really easy to change, if/when SDK guidance is different here I'll follow convention.